### PR TITLE
Reactive operations for Cosmos db

### PR DESCRIPTION
**Pull request by radovanradic**

### CODE REVIEW

**Comment by dstepanov** [comment]
Technically reactive/sync Cosmos driver is one reactive implementation that is blocking for sync code, so we don't need to choose if anything reactive should be used. I mentioned at the beginning that it would be enough to implement the reactive repository operations and for the sync to block it.

**Comment by radovanradic** [comment]
> Technically reactive/sync Cosmos driver is one reactive implementation that is blocking for sync code, so we don't need to choose if anything reactive should be used. I mentioned at the beginning that it would be enough to implement the reactive repository operations and for the sync to block it.

Ok, I thought we need somehow enable users to configure whether they use sync or reactive repo but I see now it does not make much sense. Cosmos API that we use also does the way you suggested. I have updated implementation.

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java [hunk 1]
```diff
@@ -0,0 +1,712 @@
...
+                CosmosPagedFlux<ObjectNode> result = getCosmosResults(preparedQuery, paramList, ObjectNode.class);
+                return result.map(item -> deserialize(item, argument));
+            } else {
+                DataType dataType = preparedQuery.getResultDataType();
+                Class<R> resultType = preparedQuery.getResultType();
+                CosmosPagedFlux<?> result = getCosmosResults(preparedQuery, paramList, getDataTypeClass(dataType));
+                return result.mapNotNull(item -> {
+                    if (resultType.isInstance(item)) {
+                        return (R) item;
+                    } else if (item != null) {
```
Comment (dstepanov) [inline data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java:327]: Another not needed `else`

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java [hunk 2]
```diff
@@ -0,0 +1,712 @@
...
+                        return ConversionService.SHARED.convertRequired(item, resultType);
```
Comment (dstepanov) [inline data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java:328]: Please inject `ConversionService` instance from the bean context.

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java [hunk 3]
```diff
@@ -0,0 +1,712 @@
...
+        CosmosReactiveEntitiesOperation<T> op = createCosmosReactiveBulkOperation(ctx, persistentEntity, operation, BulkOperationType.UPDATE);
+        op.update();
+        return op.getEntities();
+    }
+
+    @Override
+    @NonNull
+    public <T> Flux<T> persistAll(InsertBatchOperation<T> operation) {
+        Flux<T> result = Flux.empty();
+        for (InsertOperation<T> insertOperation : operation.split()) {
```
Comment (dstepanov) [inline data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java:381]: This should be batched in the future.

**Review by dstepanov — CHANGES_REQUESTED** [review]

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/SyncCosmosRepositoryOperations.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+import java.util.concurrent.Executors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * The sync Azure Cosmos DB operations implementation.
+ */
+@Singleton
+@Internal
+final class SyncCosmosRepositoryOperations implements
```
[the remark this answers is not in the document]
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/SyncCosmosRepositoryOperations.java:43]: Done https://github.com/micronaut-projects/micronaut-data/pull/1797/commits/8242884205f015b9a077bd3cc9a1318f31c4bdca

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java [hunk 5]
```diff
@@ -0,0 +1,712 @@
...
+        CosmosQueryRequestOptions requestOptions = new CosmosQueryRequestOptions();
+        preparedQuery.getParameterInRole(Constants.PARTITION_KEY_ROLE, PartitionKey.class).ifPresent(requestOptions::setPartitionKey);
+        return container.queryItems(querySpec, requestOptions, itemsType);
+    }
+
+    @Override
+    public <T> Mono<Boolean> exists(@NonNull PreparedQuery<T, Boolean> preparedQuery) {
+        List<SqlParameter> paramList = new ParameterBinder().bindParameters(preparedQuery);
+        CosmosPagedFlux<ObjectNode> result = getCosmosResults(preparedQuery, paramList, ObjectNode.class);
+        return result.byPage().publishOn(Schedulers.parallel())
```
[the remark this answers is not in the document]
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java]: Removed https://github.com/micronaut-projects/micronaut-data/pull/1797/commits/8242884205f015b9a077bd3cc9a1318f31c4bdca

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java [hunk 6]
```diff
@@ -0,0 +1,712 @@
...
+            if (isEntity || dtoProjection) {
+                Argument<R> argument;
+                if (dtoProjection) {
+                    argument = Argument.of(ReflectionUtils.getWrapperType(preparedQuery.getResultType()));
+                } else {
+                    argument = Argument.of(preparedQuery.getResultType());
+                }
+                CosmosPagedFlux<ObjectNode> result = getCosmosResults(preparedQuery, paramList, ObjectNode.class);
+                return result.map(item -> deserialize(item, argument));
+            } else {
```
[the remark this answers is not in the document]
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java]: Changed https://github.com/micronaut-projects/micronaut-data/pull/1797/commits/8242884205f015b9a077bd3cc9a1318f31c4bdca

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java (hunk 1 above)
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java]: Changed https://github.com/micronaut-projects/micronaut-data/pull/1797/commits/8242884205f015b9a077bd3cc9a1318f31c4bdca

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java (hunk 2 above)
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java]: Done, it was already there but just took this from other examples... Updated https://github.com/micronaut-projects/micronaut-data/pull/1797/commits/8242884205f015b9a077bd3cc9a1318f31c4bdca

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java [hunk 7]
```diff
@@ -0,0 +1,712 @@
...
+            RuntimePersistentEntity<T> persistentEntity = runtimeEntityRegistry.getEntity(type);
+            CosmosAsyncContainer container = getContainer(persistentEntity);
+            final SqlParameter param = new SqlParameter("@ROOT_ID", id.toString());
+            final SqlQuerySpec querySpec = new SqlQuerySpec(FIND_ONE_DEFAULT_QUERY, param);
+            logQuery(querySpec, Collections.singletonList(param));
+            final CosmosQueryRequestOptions options = new CosmosQueryRequestOptions();
+            if (isIdPartitionKey(persistentEntity)) {
+                options.setPartitionKey(new PartitionKey(id.toString()));
+            }
+            CosmosPagedFlux<ObjectNode> result = container.queryItems(querySpec, options, ObjectNode.class);
```
[the remark this answers is not in the document]
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java:152]: This is from their docs 
```
Cosmos implementation of ContinuablePagedFlux.
This type is a Flux that provides the ability to operate on pages of type FeedResponse and individual items in such pages. This type supports String type continuation tokens, allowing for restarting from a previously-retrieved continuation token.
```
and queryItems returns it so we have to use it. Not sure about multiple responses, but for test cases we have it worked fine.

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java (hunk 3 above)
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java]: Done now in https://github.com/micronaut-projects/micronaut-data/pull/1797/commits/45b055614910734841988465021e7fe2fd9c800a

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java [hunk 8]
```diff
@@ -0,0 +1,712 @@
...
+        try {
+            List<SqlParameter> paramList = new ParameterBinder().bindParameters(preparedQuery);
+            boolean dtoProjection = preparedQuery.isDtoProjection();
+            boolean isEntity = preparedQuery.getResultDataType() == DataType.ENTITY;
+            if (isEntity || dtoProjection) {
+                return findOneEntityOrDto(preparedQuery, paramList);
+            } else {
+                return findOneCustomResult(preparedQuery, paramList);
+            }
+        } catch (CosmosException e) {
```
[the remark this answers is not in the document]
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java]: I tried to produce 404 error, I think it was returned in sync but now it's not in reactive so I suppose we don't need to catch it. Removed

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java [hunk 9]
```diff
@@ -0,0 +1,712 @@
...
+                    ObjectNode item = iterator.next();
+                    if (iterator.hasNext()) {
+                        return Flux.error(new NonUniqueResultException());
+                    }
+                    return Mono.just(deserialize(item, Argument.of(type)));
+                }
+                return Flux.empty();
+            }).next();
+        } catch (CosmosException e) {
+            if (e.getStatusCode() == HttpResponseStatus.NOT_FOUND.code()) {
```
[the remark this answers is not in the document]
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java]: Seems like not needed to be caught explicitly here so I just removed try/catch.

**Comment by radovanradic** [comment]
@dstepanov I believe I addressed all comments.

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java [hunk 10]
```diff
@@ -0,0 +1,708 @@
...
+    @Override
+    @NonNull
+    public <T, R> Flux<R> findAll(@NonNull PreparedQuery<T, R> preparedQuery) {
+        boolean dtoProjection = preparedQuery.isDtoProjection();
+        boolean isEntity = preparedQuery.getResultDataType() == DataType.ENTITY;
+        List<SqlParameter> paramList = new ParameterBinder().bindParameters(preparedQuery);
+        if (isEntity || dtoProjection) {
+            Argument<R> argument;
+            if (dtoProjection) {
+                argument = Argument.of(ReflectionUtils.getWrapperType(preparedQuery.getResultType()));
```
Comment (dstepanov) [inline data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java:297]: Please use `getResultArgument`

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/AbstractCosmosOperations.java [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+
+/**
+ * The base class for Cosmos Db operations.
+ *
+ * @author radovanradic
+ * @since 4.0.0
+ */
+abstract class AbstractCosmosOperations extends AbstractRepositoryOperations implements
```
Comment (dstepanov) [inline data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/AbstractCosmosOperations.java:98]: We can probably remove the abstract class

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java [hunk 12]
```diff
@@ -0,0 +1,708 @@
...
+        CosmosPagedFlux<ObjectNode> result = getCosmosResults(preparedQuery, paramList, ObjectNode.class);
+        return result.byPage().flatMap(fluxResponse -> {
+            Iterator<ObjectNode> iterator = fluxResponse.getResults().iterator();
+            if (iterator.hasNext()) {
+                ObjectNode item = iterator.next();
+                if (iterator.hasNext()) {
+                    return Flux.error(new NonUniqueResultException());
+                }
+                if (preparedQuery.isDtoProjection()) {
+                    Class<R> wrapperType = ReflectionUtils.getWrapperType(preparedQuery.getResultType());
```
Comment (dstepanov) [inline data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java:212]: Please use `getResultArgument`

**Review by dstepanov — CHANGES_REQUESTED** [review]

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/AbstractCosmosOperations.java (hunk 11 above)
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/AbstractCosmosOperations.java:98]: Ugh, yes, I agree. It's just I created new branch based on this so will be difficult to merge:) But I will remove it, makes sense.

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java (hunk 12 above)
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java:212]: We agreed to leave as is for now.

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java (hunk 10 above)
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java:297]: We agreed to leave as is for now.

**Review by dstepanov — APPROVED** [review]

File: data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (radovanradic) [reply data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultReactiveCosmosRepositoryOperations.java]: Done in https://github.com/micronaut-projects/micronaut-data/pull/1797/commits/439dbd703a6abcc970b91e0ffd3124c7d93ad4c3

**Pull request merged**